pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,148 @@
+// As the create instance operation is in parallel, it is possible that an IP range was reserved by the service instance but is not updated in the cloud instance. The reserved list stores such IP ranges
+type IPAllocator struct {
+	reserved map[string]bool
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument reserved map[string]bool is a set of currently reserved IPs for which the CreateInstance operation is not completed
+func NewIPAllocator(reserved map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		reserved: reserved,
+	}
+}
+
+// Reserve function adds a particular IP block in the reserve pool
+// Argument ipBlock string is an IPV4 range which needs to be reserved
+func (ipAllocator *IPAllocator) Reserve(ipV4Range string) {
+	ipAllocator.reserved[ipV4Range] = true
+}
+
+// Unreserve function frees up a reserved IP
+// Argument ipBlock string is an IPV4 range which needs to be unreserved
+func (ipAllocator *IPAllocator) Unreserve(ipV4Range string) {
+	delete(ipAllocator.reserved, ipV4Range)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	_, _, err := net.ParseCIDR(cidr)
+	return err == nil
+}
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block.
+// cidr: Provided cidr address in which we need to look for an unreserved /29 IP Block
+// reservedIPRanges: All the Reserved IP ranges present in the cloud provider
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by merging these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	// Merging the reserved lists at the cloud provider end and the service end to form a final reserved list
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, incrementStep29IPBlock) {
+		overLap := false
+		ipRange := fmt.Sprint(cidrIPBlock.String(), reservationBlockSize)
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap, err = isOverlap(ipRange, reservedIPRange)
+			// Error while processing cidr is also considered as an overlap to ignore the specific IP block
+			if overLap || err != nil {
+				break
```

Comment:
The error is being dropped.  At this point, is it expected to receive an error? If not, then we should probably return immediately with an error

Reply:
Error is for this cidrIP and reservedIP overlap combination. We can try out other combinations.

Reply:
Yes, I mean that we should probably return if there's an error instead of continuing